compaction: fix potential data resurrection with file-based migration

nodep opened the pull request:
When tablets are migrated with file-based streaming, we can have a situation where a tombstone is garbage collected before the data it shadows lands. For instance, if we have a tablet replica with 3 sstables:

1. sstable containing an expired tombstone
2. sstable with additional data
3. sstable containing data which is shadowed by the expired tombstone in sstable 1

If this tablet is migrated, and the sstables are streamed in the order listed above, the first two sstables can be compacted before the third sstable arrives. In that case, the expired tombstone will be garbage collected, and data in the third sstable will be resurrected after it arrives to the pending replica.

This change fixes this problem by disabling tombstone garbage collection for pending replicas.

This fixes a problem in Enterprise, but the change is in OSS in order to have as few differences between OSS and Enterprise and to have a common infrastructure for disabling tombstone GC on pending replicas.

This change has to be backported to all active versions: 6.0, 6.1 and 6.2, as well as Enterprise 2024.2

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11851/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11851/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11827/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/758/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3813/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11851/)
#### Build Details:
- Duration: 3 hr 39 min
- Builder: spider8.cloudius-systems.com

raphaelsc commented:
> When tablets are migrated with file-based streaming, we can have a situation where a tombstone is garbage collected before the data it shadows lands. For instance, if we have a tablet replica with 3 sstables:
> 
> 1. sstable containing an expired tombstone
> 2. sstable with additional data
> 3. sstable containing data which is shadowed by the expired tombstone in sstable 1
> 
> If this tablet is migrated, and the sstables are streamed in the order listed above, the first two sstables can be compacted before the third sstable arrives. In that case, the expired tombstone will be garbage collected, and data in the third sstable will be resurrected after it arrives to the pending replica.
> 
> This change fixes this problem by disabling tombstone garbage collection for pending replicas.
> 
> This fixes a problem in Enterprise, but the change is in OSS in order to have as few differences between OSS and Enterprise and to have a common infrastructure for disabling tombstone GC on pending replicas. This is because of a possibility that OSS is also susceptible to a similar problem with mutation-based streaming. However, this hasn't been proven yet.
> 

I think we can have a test here too to guarantee behavior is preserved using load_and_stream.

1) start 1 node cluster.
2) set up a table with 1 tablet and gc_grace_seconds of 0.
3) create deleted data and tombstone in two different sstables (disable auto compaction).
4) stop node, move sstables to upload, resume node (see example in test_tablet_load_and_stream)
5) enable injection stream_mutation_fragments (see example in test_streaming_is_guarded_by_topology_guard)
6) add another node
7) async start migration of tablet to another node (see example in test_streaming_is_guarded_by_topology_guard)
8) waits for log 'stream_mutation_fragments: waiting'
9) run load-and-stream on resumed node, again we need to tell which sstable is streamed first. maybe we can do it in two load-and-stream rounds (see example in test_tablet_load_and_stream on how to run it).
10) signal injection stream_mutation_fragments, so migration streaming proceeds.
11) wait on async migration task.
12) verify we have no resurrection.

raphaelsc commented on replica/database.hh:
```diff
@@ -1081,6 +1081,10 @@ public:
 
+    bool tombstone_gc_enabled(locator::tablet_id tid) const {
+        return _sg_manager->tombstone_gc_enabled(tid);
+    }
```
I think we combine _tombstone_gc_enabled and _sg_manager->tombstone_gc_enabled(tid) here instead, so it's hidden from compaction_group::table_state, which will only call table::tombstone_gc_enabled(locator::tablet_id).

bhalevy commented on replica/table.cc:
```diff
@@ -993,6 +997,14 @@ tablet_storage_group_manager::maybe_split_sstable(const sstables::shared_sstable
+    if (std::find(replicas.begin(), replicas.end(), this_replica) == replicas.end()) {
+        return false;
+    }
+    return true;
```
nit: can just `return std::find(replicas.begin(), replicas.end(), this_replica) != replicas.end()`

bhalevy replied:
And how frequently is this function called?
We can precalculate this when the tablet_storage_group_manager is constructed and later in `tablet_storage_group_manager::update_effective_replication_map`.

avikivity replied:
std::ranges::contains(replicas, this_replica)

raphaelsc replied:
> And how frequently is this function called? We can precalculate this when the tablet_storage_group_manager is constructed and later in `tablet_storage_group_manager::update_effective_replication_map`.

Once for every partition (not row) that contains a tombstone. Compaction time is dominated by I/O and merging data, not iterating through a bounded replica set, so we thought it's not worth caching this info, since it's always better to have a single source of truth when possible.

raphaelsc replied:
We can measure the impact with that perf sstable tool if you'd like to, against a sstable full of expiring tombstones.

raphaelsc replied:
On a 2nd thought, maybe we don't want to waste cycles on this and the cache is worth the effort. If we go through the route of catching local tablet replicas transitioning from "pending" to "active" in update_erm(), the same mechanism can be reused for registering tables with views in the view update generator at the right time. Thoughts?

bhalevy replied:
How about small partitions?
Your assumptions could be skewed in this kind of dataset.

bhalevy replied:
I think it's worth caching the result in the storage group and updating it either incrementally and/or when the erm is updated, since those changes are relatively rare whild the per-partition check can be frequent with small partitions.

raphaelsc replied:
> I think it's worth caching the result in the storage group and updating it either incrementally and/or when the erm is updated, since those changes are relatively rare whild the per-partition check can be frequent with small partitions.

I agree. @nodep please proceed with the change. It also has the benefit of reuse for fixing the view update from staging issue.

nodep commented:
In v2:
- Added `tombstone_gc_enabled` flag to `compaction_group`. This is more efficient than checking the tablet map for every call the `table_state::tombstone_gc_enabled()`
- Added test

raphaelsc commented on replica/table.cc:
```diff
@@ -2247,7 +2247,7 @@ class compaction_group::table_state : public compaction::table_state {
     bool tombstone_gc_enabled() const noexcept override {
-        return _t._tombstone_gc_enabled;
+        return _t.tombstone_gc_enabled()  &&  _cg.tombstone_gc_enabled();
     }
```
extra space?

nodep replied:
That is my old habit of leaving an extra space around binary logical operators, which I will have to get rid of.
I don't think it is a big problem, but I can remove it here, if you insist.

raphaelsc replied:
It definitely doesn't have to block the merge, can be done later.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12137/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12137/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12137/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12021/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/945/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4002/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12137/)
#### Build Details:
- Duration: 7 hr 43 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
@tgrabiec please review and consider merging
